lib/node.js

```diff
@@ -502,6 +502,40 @@ class Node {
     return node.canReplaceWith(this)
   }
 
+  matches (node) {
+    // if the nodes are literally the same object, obviously a match.
+    if (node === this)
+      return true
+
+    // if the names don't match, they're different things, even if
+    // the package contents are identical.
+    if (node.name !== this.name)
+      return false
+
+    // if they're links, they match if the targets match
+    if (this.isLink)
+      return node.isLink && this.target.matches(node.target)
+
+    // if they're two root nodes, they're different if the paths differ
+    if (this.isRoot && node.isRoot)
+      return this.path === node.path
+
+    // if the integrity matches, then they're the same.
+    if (this.integrity && node.integrity)
+      return this.integrity === node.integrity
+
+    // if no integrity, check resolved
+    if (this.resolved && node.resolved)
+      return this.resolved === node.resolved
+
+    // if no resolved, check both package name and version
+    // otherwise, conclude that they are different things
+    return this.package.name && node.package.name &&
```

Comment:
question: when accessing `this.package` properties is there any need to be defensive about it? e.g: `this.package &&` - or is it always sure to have at least an empty object?

Reply:
It's always at least an empty object.